### PR TITLE
Remove gcp in-tree cloud provider and credential providers

### PR DESCRIPTION
**Pull request by dims**
Related to KEPs:

- https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers
- https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/541-external-credential-providers

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Removing the last remaining in-tree gcp cloud provider and credential provider. Please use the external cloud provider and credential provider from https://github.com/kubernetes/cloud-provider-gcp instead.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig cloud-provider
/sig auth

### CODE REVIEW

**Comment by dims** [comment]
/assign @xing-yang @msau42 
(for sig-storage)
/assign @elmiko 
(for sig-cloud-provider)
/assign @derekwaynecarr @SergeyKanzhelev 
(for sig-node)
/assign @liggitt @enj 
(for sig-auth)

also xref: https://github.com/kubernetes/kubernetes/issues/122828

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins_providers.go
Comment (enj) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins_providers.go:25]: These are just stubs that provide helpful error messages.  Why would we remove them?

Comment (liggitt) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins_providers.go:25]: yeah, agreed. these don't add any dependencies and are helpful pointers for people with old kubeconfig files

Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins_providers.go:25]: client-go already dropped cloud provider deps a while back: https://github.com/kubernetes/kubernetes/pull/110013/files#r871431867

**Review by elmiko — COMMENTED** [review]
i think the cloud controller changes make sense to me

/lgtm

File: pkg/credentialprovider/gcp/metadata.go
Comment (liggitt) [inline pkg/credentialprovider/gcp/metadata.go:153]: remove the `AreLegacyCloudCredentialProvidersDisabled` function and lock DisableKubeletCloudCredentialProviders to true / GA?

**Review by liggitt — COMMENTED** [review]
pkg/credentialprovider/gcp change lgtm

**Comment by mattcary** [comment]
lgtm for the storage perspective.

There are currently no tests in cloud-provider-gcp to replace the gcePersistentDiskVolumeSource tests deleted here, but there are plans to add some (https://github.com/kubernetes/kubernetes/issues/122828#issuecomment-2065367804). @BenTheElder

**Comment by dims** [comment]
> but wasn't expecting this PR just yet

yea, i start the PRs early so we can shake out things that break. for example the last crop of things we fixed were related to making sure all ci jobs are running with `cloudprovider=external` etc, making this green ensure we don't regress too ... also earlier in the cycle we land something the easier it is to fix up CI jobs etc as well. it's taken us [5 years](https://github.com/kubernetes/kubernetes/issues/76506) to get here, so will keep pushing updates whether 1.31 or 1.32 :) though i definitely hope earlier 🤞🏾

File: cmd/kube-controller-manager/app/plugins.go
Comment (carlory) [inline cmd/kube-controller-manager/app/plugins.go:121]: The portworx volume plugin doesn't complete csi migration plan, so we could not remove it from the expand-controller and pv-binder-controller in this release.

According to the previous sig-meeting, its migration plan is not clear. It lacks active maintainers and contributors to push the migration forward. So I don't know if we should remove it directly in this release without a deprecated period. What do you think about it? @xing-yang @jsafrane

File: cmd/kubelet/app/plugins_providers.go
Comment (carlory) [inline cmd/kubelet/app/plugins_providers.go:66]: I open a PR to remove the rbd. https://github.com/kubernetes/kubernetes/pull/124546

**Comment by aojea** [comment]
pkg/controller/nodeipam changes lgtm

File: cmd/kube-controller-manager/app/plugins.go
Comment (xing-yang) [reply cmd/kube-controller-manager/app/plugins.go:121]: Regarding Portworx CSI Migration, @lpabon is working on enabling it by default while staying in Beta in 1.31.

Comment (xing-yang) [reply cmd/kube-controller-manager/app/plugins.go:121]: https://github.com/kubernetes/enhancements/pull/4467

Comment (carlory) [reply cmd/kube-controller-manager/app/plugins.go:121]: @xing-yang  The KEP was merged on Feb 2. but @lpabon commented that [the plan is to remove the in-tree driver for Portworx for 1.31](https://github.com/kubernetes/kubernetes/issues/99518)

Comment (xing-yang) [reply cmd/kube-controller-manager/app/plugins.go:121]: Correct.  Please note that the milestone is 1.31 in the KEP.  @lpabon initially tried to do it in 1.30, but decided later 1.31 will be the target release.

File: cmd/kubelet/app/plugins_providers.go
Comment (dims) [reply cmd/kubelet/app/plugins_providers.go:66]: thanks!

File: cmd/kube-controller-manager/app/plugins.go
Comment (dims) [reply cmd/kube-controller-manager/app/plugins.go:121]: from what i understood ... sounds like nothing changes in this PR.

File: pkg/credentialprovider/gcp/metadata.go
Comment (dims) [reply pkg/credentialprovider/gcp/metadata.go:153]: Done!

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins_providers.go
Comment (dims) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/plugins_providers.go:25]: Ack, adding it back

**Comment by dims** [comment]
> There are currently no tests in cloud-provider-gcp to replace the gcePersistentDiskVolumeSource tests deleted here, but there are plans to add some (https://github.com/kubernetes/kubernetes/issues/122828#issuecomment-2065367804). @BenTheElder

added to the issue where we are tracking things that get dropped on the [floor](https://github.com/kubernetes/kubernetes/issues/122828#issuecomment-2088464679) to be fixed later

**Comment by dims** [comment]
@seans3 will wait for your review. it's mostly about follow ups at this point after this lands as i see it.

**Comment by carlory** [comment]
@dims This PR also removes the portworxVolume plugin from the PV controller, so the pv controller cannot provision a pv with a portworxVolume volume source once it is merged. 

FYI:

- https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/plugins.go#L121
- https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/plugins_providers.go#L62 (this pr)
- https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1550

/cc @xing-yang @lpabon 

please review again.

**Comment by dims** [comment]
@carlory please review again, i've restored the specific things you mentioned.

**Comment by carlory** [comment]
LGTM

File: test/e2e/network/scale/ingress.go
Comment (danwinship) [inline test/e2e/network/scale/ingress.go:17]: This leaves just an empty file in `test/e2e/network/scale/`. You can drop it.

Comment (dims) [reply test/e2e/network/scale/ingress.go]: Done!

**Comment by dims** [comment]
@danwinship please go ahead with your changes and i'll adapt this PR to ensure i don't drop things you need/merge. Waiting on @seans3 's initial review to figure out time line, so i don't know yet.

File: cmd/cloud-controller-manager/providers.go
Comment (andrewsykim) [inline cmd/cloud-controller-manager/providers.go:28]: Is it possible to keep this file with a dummy import or some comments? Historically `cmd/cloud-controller-manager` was used as a reference.

File: pkg/features/kube_features.go
Comment (andrewsykim) [inline pkg/features/kube_features.go:1034]: I think we also agreed to graduate this feature gate to GA in 1.31. cc @cheftako @elmiko

Comment (elmiko) [reply pkg/features/kube_features.go:1034]: yes, that was the plan in the enhancement

**Comment by jsafrane** [comment]
Please remove GCP from PV label admission plugin:
https://github.com/kubernetes/kubernetes/blob/30ae12d018697d3c5f04e225b11f242f5310e097/plugin/pkg/admission/storage/persistentvolume/label/admission.go#L202-L207

+ `findGCEPDLabels()` and anything that looks like GCE in that file.

See https://github.com/kubernetes/kubernetes/pull/124505 and https://github.com/kubernetes/kubernetes/issues/124504 for reference.

File: test/e2e/framework/providers/azure/azure.go
Comment (jsafrane) [inline test/e2e/framework/providers/azure/azure.go:1]: Not related to this PR, but I am surprised that I see this file. I thought we removed azure support in v1.30.

File: test/e2e/storage/volume_provisioning.go
Comment (jsafrane) [inline test/e2e/storage/volume_provisioning.go:114]: Please do not remove this file. Remove all cloud provider calls instead (like `checkGCEPD()` here), but keep the test - we want to make sure that Kubernetes can still provision using in-tree GCE (AzureDisk, AWS EBS, ... storage classes) using CSI migration.
The test can't check that the right volume was created (`pd-ssd` in this case), still, it can check that at least something was provisioned and it's usable in a Pod.

File: cmd/cloud-controller-manager/providers.go
Comment (dims) [reply cmd/cloud-controller-manager/providers.go:28]: Doing this now.!

File: pkg/features/kube_features.go
Comment (dims) [reply pkg/features/kube_features.go:1034]: doing this now.

**Comment by dims** [comment]
@andrewsykim @elmiko Done what you asked! please see https://github.com/kubernetes/kubernetes/pull/124519/commits/632803834ec8994f18ec75474c76d2723a95de70

@jsafrane Does this cover all the points you made? please see https://github.com/kubernetes/kubernetes/pull/124519/commits/f1c78fc35e7db4fea6ccc83b56e01a573b4c3954

File: pkg/features/kube_features.go
Comment (andrewsykim) [inline pkg/features/kube_features.go:1034]: Also update doc string https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L192-L204

File: plugin/pkg/admission/storage/persistentvolume/label/admission.go
Comment (andrewsykim) [inline plugin/pkg/admission/storage/persistentvolume/label/admission.go:201]: Maybe for a separate PR, but the Azure and vSphere provider should be removed as well.

Comment (andrewsykim) [reply plugin/pkg/admission/storage/persistentvolume/label/admission.go:201]: Actually, with those providers removed the entire admission plugin can be removed as well

Comment (andrewsykim) [reply plugin/pkg/admission/storage/persistentvolume/label/admission.go:201]: From my understanding, there should be very few users still depending on this admission controller. But for rare cases where they still need this behavior, we created an external webhook they can install instead: https://github.com/kubernetes-sigs/cloud-pv-admission-labeler 

cc @msau42 @jsafrane

File: cmd/cloud-controller-manager/providers.go
Comment (andrewsykim) [inline cmd/cloud-controller-manager/providers.go:24]: If  possible we should still reference some import so new developers know that an import to a cloud provider package is required here

File: pkg/features/kube_features.go
Comment (dims) [reply pkg/features/kube_features.go:1034]: Done!

File: plugin/pkg/admission/storage/persistentvolume/label/admission.go
Comment (dims) [reply plugin/pkg/admission/storage/persistentvolume/label/admission.go:201]: @andrewsykim being removed here - https://github.com/kubernetes/kubernetes/pull/124505

Comment (dims) [reply plugin/pkg/admission/storage/persistentvolume/label/admission.go:201]: once https://github.com/kubernetes/kubernetes/pull/124505 lands, i will remove the entire admission plugin here

**Comment by dims** [comment]
TODO(dims):
- https://github.com/kubernetes/kubernetes/pull/124519#discussion_r1589330709
- https://github.com/kubernetes/kubernetes/pull/124519#issuecomment-2089210079

**Comment by andrewsykim** [comment]
Thanks @dims! 

/lgtm

**Comment by andrewsykim** [comment]
Holding for @elmiko, @cheftako, @bridgetkromhout and others  to review as well!

**Comment by elmiko** [comment]
i've been following the discussion here and the changes make sense to me.

many thanks for @andrewsykim for adding his thoughts, i agree with his suggestions.

it seems that we broke one of the unit tests, but aside from that i'm lgtm

**Comment by dims** [comment]
> it seems that we broke one of the unit tests, but aside from that i'm lgtm

Fixed. thanks!

**Comment by dims** [comment]
/skip

`pull-kubernetes-e2e-capz-azure-file-vmss` seems unrelated

**Comment by aojea** [comment]
@dims we don't want to loose coverage on the loadbalancer tests, @danwinship has a good alternative in https://github.com/kubernetes/kubernetes/pull/124660 and we are close to one fix in cloud-provider-kind so I think next week we'll able to merge it

**Comment by dims** [comment]
> @dims we don't want to loose coverage on the loadbalancer tests, @danwinship has a good alternative in https://github.com/kubernetes/kubernetes/pull/124660 and we are close to one fix in cloud-provider-kind so I think next week we'll able to merge it

Very cool!

**Comment by jsafrane** [comment]
storage parts LGTM

**Comment by dims** [comment]
> storage parts LGTM

thanks @jsafrane !

**Comment by elmiko** [comment]
thanks @dims 

/lgtm

**Comment by elmiko** [comment]
/hold

didn't know if we are waiting for any other area reviews. feel free to remove hold if not.

**Comment by dims** [comment]
> didn't know if we are waiting for any other area reviews. feel free to remove hold if not.

Thanks @elmiko 

- second TODO was: https://github.com/kubernetes/kubernetes/pull/124519#issuecomment-2089210079
Since https://github.com/kubernetes/kubernetes/pull/124660 has not landed yet and this one removes one file that is needed there `test/e2e/network/loadbalancer.go` potentially we can re-add the file there again in that PR. I'll request @danwinship to re-add the file there.

- first TODO was: https://github.com/kubernetes/kubernetes/pull/124519#discussion_r1589330709
Since https://github.com/kubernetes/kubernetes/pull/124505 has not merged yet, so that one can be rebased and landed after this one for sure. (and may be we can remove the admission plugin entirely as well there)

I have a side note from @seans3 about making progress on moving some of the in-tree e2e tests to cloud-provider-gcp repo. Since we can always look at what we dropped in `git` and in older branches. I am happy to land this now to avoid rebasing this large of a PR again and again.

thanks everyone for the support and reviews!

/hold cancel

**Comment by carlory** [comment]
Hi @dims After this PR is merged, is it time to remove `k8s.io/legacy-cloud-providers` from `staging`?

**Comment by dims** [comment]
@carlory let's give it some time in case we find something really wrong, am sure there's quite a few things we'll have to go clean up and we'll need your help.

**Review by todaywasawesome — COMMENTED** [review]
LGTM

**Comment by sftim** [comment]
Changelog suggestion (**untested** for accuracy; this is based on my guesses as to what the current text means)
```markdown
Removed the last remaining in-tree code for integrating with GCP as cloud provider (cloud controller manager),
and code fetching GCP credentials used during container image pulls.
Please use the external cloud provider and credential provider from https://github.com/kubernetes/cloud-provider-gcp instead.
```

:information_source: It's possible the removed code was for kubectl credential exec plugins instead, I haven't checked.

**Comment by jbtk** [comment]
Hi! This change broke the autoscaler tests https://testgrid.k8s.io/sig-autoscaling-cluster-autoscaler#gci-gce-autoscaling - I started looking at them just recently (just before they broke as there were some constantly failing ones) and it seems that they might have missed some required migration. Since KEP does not mention any required migration for tests could you have a look and try pointing out what these tests are missing? They are probably missing some switch to external provider, but I am not sure where it should be made. Could you link for example migration of other tests so that I can apply the same logic to these tests?

**Comment by carlory** [comment]
This PR switches the provider implement to `NULLProvider`, it means that the provider does not do anything useful, it's there only to provide valid --provider cmdline option to allow testing of CSI migration tests of kubernetes.io/pd volume plugin.

FYI:
- https://github.com/kubernetes/kubernetes/blob/a07d3c49b301acd3426e985c3ad54a56fdb5b925/test/e2e/framework/providers/gce/gce.go#L27
- (aws) https://github.com/kubernetes/kubernetes/blob/a07d3c49b301acd3426e985c3ad54a56fdb5b925/test/e2e/framework/providers/aws/aws.go#L31-L34

**Comment by carlory** [comment]
e2e tests: https://github.com/kubernetes/kubernetes/blob/a07d3c49b301acd3426e985c3ad54a56fdb5b925/test/e2e/autoscaling/cluster_size_autoscaling.go#L99



Hi @jbtk, Do those tests must be run on a cluster with GCE, or can be run on a kind cluster? I saw that @pohly is raising a PR https://github.com/kubernetes/kubernetes/pull/123078. In that PR, it uses kind cluster to run some tests related to cluster autoscaler. 

cc @dims 

should we revert this PR or just skip e2e tests util it can use kind cluster?

- https://github.com/kubernetes/kubernetes/blob/a07d3c49b301acd3426e985c3ad54a56fdb5b925/test/e2e/autoscaling/cluster_size_autoscaling.go#L109

**Comment by carlory** [comment]
Learnt from the release notes: we may need to install `https://github.com/kubernetes/cloud-provider-gcp` before running the ginkgo test on GCP.

**Comment by aojea** [comment]
> Hi! This change broke the autoscaler tests https://testgrid.k8s.io/sig-autoscaling-cluster-autoscaler#gci-gce-autoscaling - I started looking at them just recently (just before they broke as there were some constantly failing ones) and it seems that they might have missed some required migration. Since KEP does not mention any required migration for tests could you have a look and try pointing out what these tests are missing? They are probably missing some switch to external provider, but I am not sure where it should be made. Could you link for example migration of other tests so that I can apply the same logic to these tests?

These tests that are cloud provider specific have to be moved to the corresponding cloud-provider repo @jbtk , please reach out to @seans3 and @BenTheElder 

https://github.com/kubernetes/cloud-provider-gcp/pull/683

> Learnt from the release notes: we may need to install https://github.com/kubernetes/cloud-provider-gcp before running the ginkgo test on GCP.

@carlory these tests will not work from this repo since this change

File: test/e2e/framework/providers/gce/gce.go
Comment (aojea) [inline test/e2e/framework/providers/gce/gce.go:47]: removed here the gke provider

**Review by hashim21223445 — APPROVED** [review]

**Pull request merged**
